File: optimize_images.py
"""Compress card images so the repo (and Discord) stays light.

    python3 optimize_images.py              # every image in images/
    python3 optimize_images.py a.jpg b.png  # only these files

Resizes to at most MAX_SIDE px on the longest side, re-encodes as progressive JPEG,
strips metadata. A file is only rewritten when the result is smaller.
The pre-commit hook in hooks/ runs this on staged images automatically.
"""
import io
import os
import sys

from PIL import Image, ImageOps

MAX_SIDE = 1280   # Discord shows embed images at ~1280px at most
QUALITY = 82
IMAGES_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "images")
EXTS = {".jpg", ".jpeg", ".png", ".webp"}


def optimize(path):
    before = os.path.getsize(path)
    img = ImageOps.exif_transpose(Image.open(path)).convert("RGB")
    img.thumbnail((MAX_SIDE, MAX_SIDE))
    buf = io.BytesIO()
    img.save(buf, "JPEG", quality=QUALITY, optimize=True, progressive=True)
    data = buf.getvalue()
    if len(data) < before:
        with open(path, "wb") as f:
            f.write(data)
    return before, os.path.getsize(path)


def main(files):
    if not files:
        files = [os.path.join(IMAGES_DIR, f) for f in sorted(os.listdir(IMAGES_DIR))]
    files = [f for f in files if os.path.splitext(f)[1].lower() in EXTS and os.path.isfile(f)]
    total_before = total_after = 0
    for f in files:
        b, a = optimize(f)
        total_before += b
        total_after += a
        if a < b:
            print(f"{os.path.basename(f)}: {b // 1024} KB -> {a // 1024} KB")
    if files:
        saved = 100 - total_after * 100 // max(total_before, 1)
        print(f"{len(files)} files: {total_before // 1024} KB -> {total_after // 1024} KB ({saved}% smaller)")


if __name__ == "__main__":
    main(sys.argv[1:])

File: scripts/optimize-images.ts
// Compress card images so the repo (and Discord) stays light.
//
//   bun run scripts/optimize-images.ts              # every image in images/
//   bun run scripts/optimize-images.ts a.jpg b.png  # only these files
//
// Resizes to at most MAX_SIDE px on the longest side, re-encodes as progressive JPEG, strips
// metadata. A file is only rewritten when the result is at least 5% smaller, so already
// optimized files are never re-encoded. The pre-commit hook in hooks/ runs this on staged images.
import { readdirSync, statSync, writeFileSync } from "node:fs";
import { extname, join, basename } from "node:path";
import sharp from "sharp";

const MAX_SIDE = 1280; // Discord shows embed images at ~1280px at most
const QUALITY = 82;
const IMAGES_DIR = join(import.meta.dir, "..", "images");
const EXTS = new Set([".jpg", ".jpeg", ".png", ".webp"]);

async function optimize(path: string): Promise<[number, number]> {
  const before = statSync(path).size;
  const data = await sharp(path)
    .rotate() // apply EXIF orientation, then drop the metadata
    .resize({ width: MAX_SIDE, height: MAX_SIDE, fit: "inside", withoutEnlargement: true })
    .jpeg({ quality: QUALITY, progressive: true, mozjpeg: true })
    .toBuffer();
  if (data.length < before * 0.95) writeFileSync(path, data);
  return [before, statSync(path).size];
}

let files = process.argv.slice(2);
if (!files.length) files = readdirSync(IMAGES_DIR).sort().map((f) => join(IMAGES_DIR, f));
files = files.filter((f) => EXTS.has(extname(f).toLowerCase()) && statSync(f).isFile());

let totalBefore = 0, totalAfter = 0;
for (const f of files) {
  const [b, a] = await optimize(f);
  totalBefore += b;
  totalAfter += a;
  if (a < b) console.log(`${basename(f)}: ${b >> 10} KB -> ${a >> 10} KB`);
}
if (files.length) {
  const saved = 100 - Math.floor((totalAfter * 100) / Math.max(totalBefore, 1));
  console.log(`${files.length} files: ${totalBefore >> 10} KB -> ${totalAfter >> 10} KB (${saved}% smaller)`);
}
